pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -150,10 +158,74 @@ func (t *TrainJob) PodLabelSelector() string {
 }
 
 func getChildJobSet(t *TrainJob) (*jobsetapi.JobSet, error) {
-	jobSet := jobsetapi.JobSet{}
-	// TODO: Ideally we should be using the parent context here
-	err := reconciler.client.Get(context.Background(), types.NamespacedName{Name: t.Name, Namespace: t.Namespace}, &jobSet)
-	return &jobSet, err
+	runtimeRefGK := kftrainerruntime.RuntimeRefToRuntimeRegistryKey(t.Spec.RuntimeRef)
+	runtime, ok := reconciler.runtimes[runtimeRefGK]
+	if !ok {
+		return nil, fmt.Errorf("unsupported runtime: %s", runtimeRefGK)
+	}
+
+	trainJob := (*kftrainerapi.TrainJob)(t)
+	trSpec, err := getRuntimeSpec(trainJob)
+	if err != nil {
+		return nil, fmt.Errorf("runtime '%s' not found", trainJob.Spec.RuntimeRef.Name)
+	}
+	info, err := runtime.RuntimeInfo(trainJob, trSpec.Template, trSpec.MLPolicy, trSpec.PodGroupPolicy)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the jobsetSpecApply and apply the TrainJob object overrides for the trainer and initializer jobs
+	jobSetSpec, ok := kftrainerruntime.TemplateSpecApply[jobsetapplyapi.JobSetSpecApplyConfiguration](info)
+	if !ok {
+		return nil, err
+	}
+	jobsetApply := kftrainerjobset.NewBuilder(jobsetapplyapi.JobSet(t.Name, t.Namespace).
+		WithSpec(jobSetSpec)).Initializer(trainJob).Trainer(info, trainJob).PodLabels(info.Scheduler.PodLabels).Build()
+
+	// convert to jobset with the defaults set
+	return jobsetApplyToJobset(jobsetApply)
+}
+
+func jobsetApplyToJobset(jobsetApply *jobsetapplyapi.JobSetApplyConfiguration) (*jobsetapi.JobSet, error) {
+	jsonData, err := json.Marshal(jobsetApply)
+	if err != nil {
+		return nil, err
+	}
+
+	jobset := &jobsetapi.JobSet{}
+	if err := json.Unmarshal(jsonData, jobset); err != nil {
+		return nil, err
+	}
+
+	// Run a dry-run patch to set the jobset defaults
+	// Defaults must be applied here because Kueue later compares podsets to match workloads.
+	// Workloads coming from the API server are already defaulted, so without defaulting this JobSet, matching would fail.
+	if err = reconciler.client.Patch(context.Background(), jobset, client.Apply, &client.PatchOptions{
+		FieldManager: "defaulter",
+		DryRun:       []string{metav1.DryRunAll},
+	}); err != nil {
+		return nil, err
+	}
+	return jobset, nil
+}
+
+func getRuntimeSpec(trainJob *kftrainerapi.TrainJob) (*kftrainerapi.TrainingRuntimeSpec, error) {
+	// Defaults to ClusterTrainingRuntime according to trainjob_types.go
+	if trainJob.Spec.RuntimeRef.Kind == nil || *trainJob.Spec.RuntimeRef.Kind == "ClusterTrainingRuntime" {
```

Comment:
Do we run the Kueue TrainJob controller after the TrainJob defaulter is running ?
As you can see, we set the default value here: https://github.com/kubeflow/trainer/blob/master/pkg/apis/trainer/v1alpha1/trainjob_types.go#L151